### ISSUE TITLE
Redirect URI should not be required for Confidential Client

### ISSUE THREAD

**Issue by onovotny** [AzureAD/microsoft-authentication-library-for-dotnet#426]
Some flows, like the On-behalf-of flow don't require a redirect URI, so that value shouldn't be required in the ConfidentialClient constructor.

**Comment by kpanwar** [comment]
You should be able to pass null and not run into issues.

**Comment by kpanwar** [comment]
Sure! I will kee this issue open to track the work.

**Comment by htuomola** [comment]
Actually passing redirect uri as null causes an exception when calling `AcquireTokenOnBehalfOfAsync(scopes, UserAssertion)`, on creating request params. 
```
System.ArgumentNullException : Value cannot be null.
Parameter name: uriString
   at System.Uri..ctor(String uriString)
   at Microsoft.Identity.Client.ClientApplicationBase.CreateRequestParameters(Authority authority, IEnumerable`1 scopes, IUser user, TokenCache cache)
   at Microsoft.Identity.Client.ConfidentialClientApplication.CreateRequestParameters(Authority authority, IEnumerable`1 scopes, IUser user, TokenCache cache)
   at Microsoft.Identity.Client.ConfidentialClientApplication.<AcquireTokenOnBehalfCommonAsync>d__16.MoveNext()
```

I think it comes down to this, which won't work unless `RedirectUri` defaults to some valid uri:
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/d5026d5b54e845ed9f3b251c8b3c59391b49e115/src/Microsoft.Identity.Client/ClientApplicationBase.cs#L234

Code itself: 
```csharp
var scopes = new[] {"https://graph.microsoft.com/user.read"};
var userTokenCache = new TokenCache();
var cca = new ConfidentialClientApplication(clientSettings.ClientId, null, new ClientCredential(clientSettings.AppKey), userTokenCache, null);
await cca.AcquireTokenOnBehalfOfAsync(scopes, GetUserAssertion(userInformation))
```

**Comment by jmprieur** [comment]
We need a redirect URL for the authorization code grant, but not for client creds, OBO, ...
The idea is to make it optional by adding overrides, or accept null and throw when using the authorization code throw.

**Comment by jmprieur** [comment]
@bgavrilMS : this one is poachable

**Comment by bgavrilMS** [comment]
@jmprieur - I propose that we make sure the docs emphasize where redirect URI can be null and fix the exception for OBO flow and client creds. 

Alternatively, I can obsolete the calls requiring a redirectUri so that we can remove them in when we get rid of the -preview tag?

**Comment by jmprieur** [comment]
@bgavrilMS : I'd prefer the first option (allow null, fix the exception for Client credentials. Not sure for OBO?)
The alternative would touch the constructor … and most confidential client flow require the redirect URI. I would think OBO does?

**Comment by MSAppsDev** [comment]
Hi,
Is this fixed?

**Comment by bgavrilMS** [comment]
@MSAppsDev - no, the issue is still open.

**Comment by jmprieur** [comment]
@MarkZuber : done part of the confidential client builder based app configuration

**Comment by MarkZuber** [comment]
In 3.x, constructing with null will set the default redirecturi to be urn:ietf:wg:oauth:2.0:oob and that redirecturi will get used during any requests as needed.

I've added unit tests to explicitly cover this case in dev3x.

**Comment by jennyf19** [comment]
@MarkZuber just for confidential clients?

**Comment by jmprieur** [comment]
yes, just for Client Credentials, actually (daemon apps)

**Comment by jennyf19** [comment]
Fixed in [MSAL 3.0.0-preview release](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/releases/tag/3.0.0-preview)
